Treat Spark bucketed tables as unbucketed

MiguelWeezardo opened the pull request:
This is a workaround for [SPARK-19256](https://issues.apache.org/jira/browse/SPARK-19256)

hashhar commented on plugin/trino-hive/src/test/java/io/trino/plugin/hive/AbstractTestHive.java:
```diff
@@ -1799,6 +1799,27 @@ private void doTestBucketedTableValidation(HiveStorageFormat storageFormat, Sche
+            transaction.commit();
+        }
+
+        // read succeeds on tables marked with Spark-specific property
```
Can we add another test for a bucketed table written by Spark but with the table properties removed and then asserting that the reads fail? Or is there such case already covered in the sampling bucketing validator cursor?

MiguelWeezardo replied:
Sure, I have a Spark-generated TPCH nation table in ORC format somewhere. I'll see how easy it is to add it the tests.

MiguelWeezardo replied:
Turns out this is more difficult than I originally estimated. The `AbstractTestHive` framework only allows creating empty tables and injecting rows from a `MaterializedResult` with Presto, I don't see any way to import an existing set of bucketed ORC files. I've also considered writing a product test, with an actual Spark instance creating a Hive-compatible table, but it turned out Spark doesn't support bucketing in this mode.

findepi replied:
In some tests, we copy a prepared file into table's location.
see eg https://github.com/trinodb/trino/blob/072ac47331d96a8bee79bc83a971b11406ca69dc/plugin/trino-hive/src/test/java/io/trino/plugin/hive/TestHiveIntegrationSmokeTest.java#L3573

hashhar replied:
Note that you might need a local Spark instance anyway to prepare the test files (one-time activity) by having Spark write a bucketed dataframe into ORC files (no need to write a Hive table).

Also this better captures the actual workflow people have when they face issues querying Spark bucketed tables from Trino. They use Spark to write files into a location which happens to be a Hive table location.

findepi replied:
> They use Spark to write files into a location which happens to be a Hive table location.


in such a case current fix will not kick in, will it?

hashhar replied:
> in such a case current fix will not kick in, will it?

Yes. No way to know if files were bucketed. But that can happen without Spark in the picture too - with people copying files into a bucketed table without understanding the file naming nomenclature.

findepi replied:
I agree. And we would need to ask user for help (aka session property) in such a case.
Then the question is -- which problem are we trying to solve here?

hashhar replied:
Spark does allow writing to bucketed tables - IIRC newer (very recent) versions prevent writing to a bucketed Hive table but it can be overridden.

MiguelWeezardo replied:
> Note that you might need a local Spark instance anyway to prepare the test files (one-time activity) by having Spark write a bucketed dataframe into ORC files (no need to write a Hive table).

I was thinking about preparing those ORC files once and adding them to resources. That way I'd avoid running Spark.

MiguelWeezardo replied:
Turns out the initial approach with `AbstractTestHive` was the right way to go, I was just missing a couple of pieces. Thanks for your suggestions.

hashhar commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/util/HiveBucketing.java:
```diff
@@ -180,6 +181,10 @@ static int getBucketNumber(int hashCode, int bucketCount)
 
     public static Optional<HiveBucketHandle> getHiveBucketHandle(ConnectorSession session, Table table, TypeManager typeManager)
     {
+        if (table.getParameters().containsKey(SPARK_TABLE_PROVIDER_KEY)) {
```
Does this change also hide the hidden `$bucket` columns etc.?

findepi commented:
There are many different scenarios:

- Spark-created table
- Spark-created files registered in metastore by someone else (eg Trino CREATE TABLE with external_location)
- Spark inserting to an existing bucketed table
- per https://github.com/trinodb/trino/pull/6848#discussion_r573829894, a Spark version may also be a dimension (@hashhar do you have a reference?)

@MiguelWeezardo please enhance commit message (and PR description) to say what scenario is being covered and what is not covered.

hashhar commented:
Yes, since Spark 2.3 they added some safeguards in place to prevent Spark from INSERTing into Hive bucketed tables without explicitly configuring Spark to do so.

See https://github.com/apache/spark/commit/d2416925c439b308b3e903171bd28d8cfc2bf9cf for more details (the commit message explains the changes very well thankfully).

MiguelWeezardo commented:
> Yes, since Spark 2.3 they added some safeguards in place to prevent Spark from INSERTing into Hive bucketed tables without explicitly configuring Spark to do so.
> 
> See [apache/spark@d241692](https://github.com/apache/spark/commit/d2416925c439b308b3e903171bd28d8cfc2bf9cf) for more details (the commit message explains the changes very well thankfully).

I should have expressed myself on this point more clearly, sorry for the misunderstanding. 

Yes, you can force Spark to write bucketed data to Hive. The problem I encountered while writing product tests was actually different - I tried to create a Hive-format table using Spark, and it complained that `HIVE` is not a valid in `CREATE TABLE ... USING HIVE ...`. It may be explained by [documentation on Hive-formatted tables](https://spark.apache.org/docs/latest/sql-data-sources-hive-tables.html#specifying-storage-format-for-hive-tables): `Note that, Hive storage handler is not supported yet when creating table, you can create a table using storage handler at Hive side, and use Spark SQL to read it`.

> From what I understand, this PR will help with (because Spark adds it's own table properties after any operation that modifies the table - insert, schema change, partition addition):
> 
>     * Spark-created table
> 
>     * Spark inserting to an existing bucketed table
> 
> 
> But will not help with:
> 
>     * Spark-created files registered in metastore by someone else (eg Trino CREATE TABLE with external_location)
> 
> 
> And Spark version doesn't matter except for the users of Spark who might try to write to a bucketed table.

The test currently implemented in this PR takes Spark-bucketed ORC files and creates a Trino external table from them. I'm not sure if this what our customers do, but we got that one covered. I'll be looking for ways to test other cases, `Spark inserting to an existing bucketed table` sounds possible, but `Spark creating a Hive-compatible table` doesn't for the reasons explained above.

MiguelWeezardo commented:
I'm converting this to a draft until I get the tests fixed.

MiguelWeezardo commented:
I've added a product-level test for `Spark inserting to an existing bucketed table` but discovered that the table wasn't marked with the `spark.sql.sources.provider` parameter after Spark insert.

MiguelWeezardo commented:
I've added a product-level test for `Spark inserting to an existing bucketed table` but discovered that the table wasn't marked with the `spark.sql.sources.provider` parameter after Spark insert.

hashhar commented:
> I've added a product-level test for Spark inserting to an existing bucketed table but discovered that the table wasn't marked with the spark.sql.sources.provider parameter after Spark insert.

Any other `spark.sql` prefixed parameter? From what I can see the provider parameter is added for "data source" tables. (ref: https://github.com/apache/spark/blob/master/sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala#L308)

data source schema parameter (ref: https://github.com/apache/spark/blob/master/sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala#L1333) is one that must be present because Spark doesn't use the metastore as source of truth for schema (because metastore is case-insensitive).

MiguelWeezardo commented:
> > I've added a product-level test for Spark inserting to an existing bucketed table but discovered that the table wasn't marked with the spark.sql.sources.provider parameter after Spark insert.
> 
> Any other `spark.sql` prefixed parameter? From what I can see the provider parameter is added for "data source" tables. (ref: https://github.com/apache/spark/blob/master/sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala#L308)
> 
> data source schema parameter (ref: https://github.com/apache/spark/blob/master/sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala#L1333) is one that must be present because Spark doesn't use the metastore as source of truth for schema (because metastore is case-insensitive).

It seems that [Hive-compliant tables are not "data source" tables](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala#L849). I've also added a test with a Hive-compliant table created by Spark and it also doesn't have `provider` or `schema` parameters. I figured out a way to add those properties manually.

MiguelWeezardo commented:
@findepi If you think this version is OK, please merge.

findepi commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/BackgroundHiveSplitLoader.java:
```diff
@@ -137,7 +137,9 @@
-            BUCKET_WITH_OPTIONAL_ATTEMPT_ID_PATTERN);
+            BUCKET_WITH_OPTIONAL_ATTEMPT_ID_PATTERN,
+            // Spark-generated file, eg. part-00000-6666e260-fda8-4df5-a412-0a07a90b35a5_00000.c000.snappy.orc
+            Pattern.compile("part-(\\d{5})-[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}.*"));
```
This sounds incorrect. We should not attempt to derived bucket number from a file path created by Spark, since the bucket number is meaningless to us. Why do you need this?

MiguelWeezardo replied:
It looks like the tables created by Spark in both native and Hive-compliant modes are marked as transactional, and Hive connector expects there to be a bucket number on all "original" files:
```                
// Hive requires "original" files of transactional tables to conform to the bucketed tables naming pattern, to match them with delete deltas.
int bucketId = getRequiredBucketNumber(originalFilePath);
```
Here's the stacktrace:
```
tests               | 	Caused by: java.lang.IllegalStateException: Cannot get bucket number from path: hdfs://hadoop-master:9000/user/hive/warehouse/test_presto_reading_spark_buckets/part-00000-cbacc54d-5b1d-4b57-a82f-1f0ca2728de0-c000.snappy.orc
tests               | 		at io.trino.plugin.hive.BackgroundHiveSplitLoader.lambda$getRequiredBucketNumber$11(BackgroundHiveSplitLoader.java:859)
tests               | 		at java.base/java.util.OptionalInt.orElseThrow(OptionalInt.java:271)
tests               | 		at io.trino.plugin.hive.BackgroundHiveSplitLoader.getRequiredBucketNumber(BackgroundHiveSplitLoader.java:859)
tests               | 		at io.trino.plugin.hive.BackgroundHiveSplitLoader.loadPartition(BackgroundHiveSplitLoader.java:533)
tests               | 		at io.trino.plugin.hive.BackgroundHiveSplitLoader.loadSplits(BackgroundHiveSplitLoader.java:337)
tests               | 		at io.trino.plugin.hive.BackgroundHiveSplitLoader$HiveSplitLoaderTask.process(BackgroundHiveSplitLoader.java:266)
tests               | 		... 6 more
```
The files created by Spark have a bucket number contained in the filename, but none of the patterns match them. Do you think I should rather look for a way to disable creation of transactional tables by Spark?

MiguelWeezardo replied:
Turns out disabling transactions is quite easy, we can go that route.

findepi replied:
> tables created by Spark in both native and Hive-compliant modes are marked as transactional


that's weird. Last time i checked, Spark did not support transactional tables natively.

MiguelWeezardo replied:
Maybe it's a feature of Spark 3? I found some comments about Spark not supporting transactions but it was for Spark 2.x.

findepi commented on testing/trino-product-tests/src/main/java/io/trino/tests/hive/TestSparkCompatibility.java:
```diff
@@ -0,0 +1,175 @@
+        assertThat(sparkSelect).containsOnly(row1, row2, row3, row4);
+
+        assertThat(() -> onPresto().executeQuery(format("%s FROM %s", startOfSelect, prestoTableName(baseTableName))))
+                .failsWithMessage("Hive table is corrupt.");
```
I.e. the table is read with bucketing. Add a code comment here that this is expected at this stage of table definition, plus extend the exception message to be more like `Hive table is corrupt. File '%s' is for bucket %s, but contains a row for bucket %s.`

findepi commented on testing/trino-product-tests/src/main/java/io/trino/tests/hive/TestSparkCompatibility.java:
```diff
@@ -0,0 +1,175 @@
+        onHive().executeQuery(format("ALTER TABLE %s SET TBLPROPERTIES ('%s' = 'hive')", baseTableName, SPARK_TABLE_PROVIDER_KEY));
+
+        QueryResult prestoSelect = onPresto().executeQuery(format("%s FROM %s", startOfSelect, prestoTableName(baseTableName)));
+        assertThat(prestoSelect).containsOnly(row1, row2, row3, row4);
```
Add a test that selecting `$bucket` or filtering on `$bucket = 0` fail with column not found

findepi commented on testing/trino-product-tests/src/main/java/io/trino/tests/hive/TestSparkCompatibility.java:
```diff
@@ -0,0 +1,175 @@
+    }
+
+    @Test(groups = {HIVE_SPARK_BUCKETING, PROFILE_SPECIFIC_TESTS})
+    public void testPrestoReadingTableCreatedByHiveCompliantSpark()
```
Add some commentary what "native spark", "hive compliant spark" mean.

Also, if "hive compliant spark" cannot reproduce the situation without our help (no bucketing support, doesn't set the required table properties), i am not sure if this test models a real-life scenario, does it?

MiguelWeezardo replied:
I'm not sure about the real-life scenario argument - it's quite possible that a user will have a Hive-compliant table created in Spark, which up until was also accessed by Hive, and would like to read it through Trino. I think it's better to have a tested scenario for that case, if only to know what to do.

MiguelWeezardo replied:
I forgot that this table wouldn't have bucketing applied by Spark. I agree this scenario seems "hacky", with bucketing and Spark table property being added manually. Would you like to remove it?

findepi commented on testing/trino-product-tests/src/main/java/io/trino/tests/hive/TestSparkCompatibility.java:
```diff
@@ -0,0 +1,175 @@
+    public void testPrestoReadingDataInsertedBySpark()
+    {
+        String baseTableName = "presto_reading_data_inserted_by_spark";
+        String prestoTableDefinition =
```
here and in method names -- let's use "trino" (even though the test code uses `onPresto`). less work later.

findepi replied:
bump

findepi commented on testing/trino-product-tests/src/main/java/io/trino/tests/hive/TestSparkCompatibility.java:
```diff
@@ -0,0 +1,175 @@
+
+        String sparkTableDefinition =
+                "CREATE TABLE %s (" +
+                        "  string_col STRING" +
```
above you used _string, but i think a_string would look better.
the `_` prefix looks like it could have some special meaning, which is not intended here.

hashhar reviewed (commented):
Looks good to me % existing comments and please add a test for `$bucket` column.

MiguelWeezardo commented:
> Looks good to me % existing comments and please add a test for `$bucket` column.

Sure, I will add a query filtering on `$bucket`

MiguelWeezardo commented:
I found the test where Spark inserts data into a table created by Presto keeps failing the PR, but passes on my laptop. The problem is the expected error message differs - instead of reporting an error like `Hive table is corrupt. File '...' is for bucket X, but contains a row for bucket Y.` it reports an invalid number of files in the directory:
```
Caused by: io.trino.spi.TrinoException: Hive table 'default.trino_reading_data_inserted_by_spark' is corrupt. File 'part-00000-4f5648b4-b94d-44df-b9cb-fa71b89b9285-c000.snappy.orc' does not match the standard naming pattern, and the number of files in the directory (2) does not match the declared bucket count (4) for partition: <UNPARTITIONED>
at io.trino.plugin.hive.BackgroundHiveSplitLoader.getBucketedSplits(BackgroundHiveSplitLoader.java:770)
at io.trino.plugin.hive.BackgroundHiveSplitLoader.loadPartition(BackgroundHiveSplitLoader.java:558)
at io.trino.plugin.hive.BackgroundHiveSplitLoader.loadSplits(BackgroundHiveSplitLoader.java:337)
at io.trino.plugin.hive.BackgroundHiveSplitLoader$HiveSplitLoaderTask.process(BackgroundHiveSplitLoader.java:266)
at io.trino.plugin.hive.util.ResumableTasks$1.run(ResumableTasks.java:38)
at io.trino.$gen.Trino_ef6dd1e____20210221_064424_2.run(Unknown Source)
at io.airlift.concurrent.BoundedExecutor.drainQueue(BoundedExecutor.java:80)
at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
at java.base/java.lang.Thread.run(Thread.java:834)
```

findepi commented:
@electrum PTAL if you find time

findepi commented on plugin/trino-hive/src/test/java/io/trino/plugin/hive/AbstractTestHive.java:
```diff
@@ -1806,6 +1806,27 @@ private void doTestBucketedTableValidation(HiveStorageFormat storageFormat, Sche
                     .hasMessageMatching("Hive table is corrupt\\. File '.*/000002_0_.*' is for bucket 2, but contains a row for bucket 5.");
         }
+
+        // mark table as created by Spark
```
split out into a new test

findepi commented on plugin/trino-hive/src/test/java/io/trino/plugin/hive/AbstractTestHiveLocal.java:
```diff
@@ -126,4 +161,139 @@ public void testGetTableSchemaOffline()
+                            ImmutableList.of(new SortingColumn("name", SortingColumn.Order.ASCENDING)))),
+                    new Path(URI.create("file://" + externalLocation.toString())));
+
+            // read fails due to validation failure
```
Why overridden ?

MiguelWeezardo replied:
What's overridden here? This is a new test with a different name, it doesn't override anything from super.

findepi commented on plugin/trino-hive/src/test/java/io/trino/plugin/hive/AbstractTestHiveLocal.java:
```diff
@@ -126,4 +161,139 @@ public void testGetTableSchemaOffline()
+        }
+    }
+
+    private java.nio.file.Path extractResourceZipToTemporaryDirectory(String resourceName)
```
Looks like quite some amount of code being copied from super, why?

MiguelWeezardo replied:
I can add some helper methods to avoid this.

findepi commented on plugin/trino-hive/src/test/java/io/trino/plugin/hive/AbstractTestHiveLocal.java:
```diff
@@ -126,4 +161,139 @@ public void testGetTableSchemaOffline()
+            }
+        }
+        return tempDir;
+    }
```
What's the `spark_bucketed_nation.zip` being added in this PR? looks empty

MiguelWeezardo replied:
Looks like a mistake, I'll check.

MiguelWeezardo replied:
It looks like GitHub is reporting the wrong file size for `spark_bucketed_nation.zip`, the actual file is 4609 bytes, and contains a zipped directory with Spark-generated buckets.

findepi replied:
Why would we want to use a zip for these?

electrum replied:
Agreed, can we simply add them as resources?

MiguelWeezardo replied:
In one of the earlier iterations I've tried running the test with external table location pointing to a resource dir. It worked fine locally when I had sources checked out, but failed on CI reporting directory not found. The path pointed inside a JAR file. I could change the ZIP file into a resource dir, but I'd still need to extract the files from the JAR to some temporary directory, and this seems easier when just one ZIP file is involved.

electrum replied:
I think it's cleaner to explicitly copy the resources to temporary files, rather than unzip a zip file, since that makes it clear what is going on (otherwise you need to go look in the zip file).

MiguelWeezardo replied:
> I think it's cleaner to explicitly copy the resources to temporary files, rather than unzip a zip file, since that makes it clear what is going on (otherwise you need to go look in the zip file).

Sure, I changed the setup to use a resource directory instead of a zip file.

findepi commented on testing/trino-product-tests-launcher/src/main/java/io/trino/tests/product/launcher/env/environment/SinglenodeSparkHive.java:
```diff
@@ -0,0 +1,98 @@
+    @SuppressWarnings("resource")
+    private DockerContainer createSpark()
+    {
+        DockerContainer container = new DockerContainer("ghcr.io/trinodb/testing/spark3.0-iceberg:" + hadoopImagesVersion, "spark")
```
We should have Spark container w/o Iceberg. Pease file an issue and add TODO comment here.

findepi commented on testing/trino-product-tests-launcher/src/main/resources/docker/presto-product-tests/conf/environment/singlenode-spark-hive/hive.properties:
```diff
@@ -0,0 +1,15 @@
+connector.name=hive-hadoop2
+hive.metastore.uri=thrift://hadoop-master:9083
+hive.config.resources=/docker/presto-product-tests/conf/presto/etc/hive-default-fs-site.xml
+hive.allow-add-column=true
+hive.allow-drop-column=true
+hive.allow-rename-column=true
+hive.allow-comment-table=true
+hive.allow-comment-column=true
+hive.allow-drop-table=true
+hive.allow-rename-table=true
+hive.allow-register-partition-procedure=true
+hive.metastore-cache-ttl=0s
+hive.fs.cache.max-size=10
+hive.max-partitions-per-scan=100
+hive.translate-hive-views=true
```
Which of these are actually relevant, if any?

MiguelWeezardo replied:
I think only drop-table is needed. I'll trim it down.

MiguelWeezardo replied:
Turns out drop-table was for one of the "hacky" tests I already removed, so I left a simple hive connector config.

findepi commented on testing/trino-product-tests/src/main/java/io/trino/tests/hive/TestSparkCompatibility.java:
```diff
@@ -0,0 +1,87 @@
+
+        assertThat(() -> onPresto().executeQuery(format("%s, \"$bucket\" FROM %s", startOfSelect, trinoTableName(baseTableName))))
+                .failsWithMessage("Column '$bucket' cannot be resolved");
+
```
Add a SHOW CREATE TABLE test and verify presented properties exactly.

findepi commented on testing/trino-product-tests/src/main/java/io/trino/tests/hive/TestSparkCompatibility.java:
```diff
@@ -0,0 +1,87 @@
+        onSpark().executeQuery("DROP TABLE " + baseTableName);
+    }
+
+    private static String trinoTableName(String tableName)
```
inline

MiguelWeezardo commented:
@electrum, please merge if you think this is OK.

MiguelWeezardo commented:
@findepi It seems @electrum is busy with other tasks. Would it be possible to merge this?

hashhar commented on testing/trino-product-tests/src/main/java/io/trino/tests/hive/TestSparkCompatibility.java:
```diff
@@ -0,0 +1,104 @@
+        QueryResult trinoSelect = onPresto().executeQuery(format("%s FROM %s", startOfSelect, format("%s.default.%s", TRINO_CATALOG, baseTableName)));
+        assertThat(trinoSelect).containsOnly(row1, row2, row3, row4);
+
+        assertThat(onSpark().executeQuery(format("SHOW CREATE TABLE %s", baseTableName)))
```
What is this trying to test?

MiguelWeezardo replied:
It was requested in https://github.com/trinodb/trino/pull/6848#discussion_r583656065
I guess it checks the hidden columns are actually hidden.

findepi replied:
https://github.com/trinodb/trino/pull/6848#discussion_r583656065 did not request testing `SHOW CREATE TABLE` on _spark_. 

Please remove, as you have SHOW CREATE TABLE test with Trino below

hashhar reviewed (approved):
LGTM % comments.

Also, please rebase since the PR is very old now to re-run CI.
Also, have you verified that the new tests indeed fail without your patch applied?

electrum reviewed (commented):
Overall looks good to me. I'll defer to @findepi for final approval since he's more familiar with the product tests

MiguelWeezardo commented:
> Why would we want to use a zip for these?

In one of the earlier iterations I've tried running the test with external table location pointing to a resource dir. It worked fine locally when I had sources checked out, but failed on CI reporting directory not found. The path on CI pointed inside a JAR file, while in my local tests it pointed to the resource dir in `target`. I could turn the ZIP file into a resource dir, but I'd still need to extract the files from the JAR to some temporary directory for CI to pass, and this seems easier when just one ZIP file is involved.

MiguelWeezardo commented:
@findepi Please merge if you think this version is OK.

findepi commented on plugin/trino-hive/src/test/java/io/trino/plugin/hive/AbstractTestHiveLocal.java:
```diff
@@ -126,4 +159,129 @@ public void testGetTableSchemaOffline()
+                            BUCKETING_V1,
+                            3,
+                            ImmutableList.of(new SortingColumn("name", SortingColumn.Order.ASCENDING)))),
+                    new Path(URI.create("file://" + externalLocation.toString())));
```
what happens if you replace this with 
```
new Path(Resources.getResource("spark_bucketed_nation").toURI())
```

?

(using `com.google.common.io.Resources`)

findepi replied:
Hm.... weird. We load resouces from src/test/resources as _files_ relatively often in tests

maybe it was iniitally in a different module?

findepi replied:
Oh, the test is actually run from `trino-hive-hadoop2`, right?

findepi replied:
I see now we this will fail. You can revert `Use resource directory directly instead of copying to tempDir` commit, no need for CI.

findepi commented on testing/trino-product-tests-launcher/src/main/java/io/trino/tests/product/launcher/env/environment/SinglenodeSparkHive.java:
```diff
@@ -0,0 +1,99 @@
+    public void extendEnvironment(Environment.Builder builder)
+    {
+        builder.configureContainer(HADOOP, dockerContainer -> {
+            dockerContainer.setDockerImageName("ghcr.io/trinodb/testing/hdp3.1-hive:" + hadoopImagesVersion);
```
What's the reason here to use HDP3 image instead of the default one?

findepi commented on testing/trino-product-tests-launcher/src/main/java/io/trino/tests/product/launcher/env/environment/SinglenodeSparkHive.java:
```diff
@@ -0,0 +1,99 @@
+            dockerContainer.setDockerImageName("ghcr.io/trinodb/testing/hdp3.1-hive:" + hadoopImagesVersion);
+            dockerContainer.withCopyFileToContainer(
+                    forHostPath(dockerFiles.getDockerFilesHostPath("conf/environment/singlenode-hdp3/apply-hdp3-config.sh")),
+                    CONTAINER_HADOOP_INIT_D + "apply-hdp3-config.sh");
```
please remove if you switch to using the default image

findepi commented on testing/trino-product-tests-launcher/src/main/java/io/trino/tests/product/launcher/env/environment/SinglenodeSparkHive.java:
```diff
@@ -0,0 +1,99 @@
+    @SuppressWarnings("resource")
+    private DockerContainer createSpark()
+    {
+        // TODO: Switch to pure Spark 3 image once it's available (https://github.com/trinodb/trino/issues/7063)
```
https://github.com/trinodb/trino/issues/7063

findepi commented:
https://github.com/trinodb/trino/pull/6848#discussion_r590262430 + squash please.

MiguelWeezardo commented:
I've tested with default HDP version, but got this failure from Spark while creating the table (the `transactional` table property was removed):
```
spark               | 21/03/09 12:04:02 INFO SparkExecuteStatementOperation: Submitting query 'CREATE TABLE `default`.`test_trino_reading_spark_native_buckets_1uwewxkiauvk` (
spark               |   `a_string` STRING,
spark               |   `a_bigint` BIGINT,
spark               |   `an_integer` INT,
spark               |   `a_real` FLOAT,
spark               |   `a_double` DOUBLE,
spark               |   `a_boolean` BOOLEAN)
spark               | USING ORC
spark               | CLUSTERED BY (a_string)
spark               | INTO 4 BUCKETS' with 8e032443-252c-49d4-ba6d-a6fb63cf37ec
spark               | 21/03/09 12:04:02 INFO SparkExecuteStatementOperation: Running query with 8e032443-252c-49d4-ba6d-a6fb63cf37ec
spark               | 21/03/09 12:04:02 INFO metastore: Closed a connection to metastore, current connections: 0
spark               | 21/03/09 12:04:02 INFO metastore: Trying to connect to metastore with URI thrift://hadoop-master:9083
spark               | 21/03/09 12:04:02 INFO metastore: Opened a connection to metastore, current connections: 1
spark               | 21/03/09 12:04:02 INFO metastore: Connected to metastore.
spark               | 21/03/09 12:04:02 INFO DAGScheduler: Asked to cancel job group 8e032443-252c-49d4-ba6d-a6fb63cf37ec
spark               | 21/03/09 12:04:02 ERROR SparkExecuteStatementOperation: Error executing query with 8e032443-252c-49d4-ba6d-a6fb63cf37ec, currentState RUNNING,
spark               | org.apache.spark.sql.AnalysisException: org.apache.hadoop.hive.ql.metadata.HiveException: Unable to fetch table test_trino_reading_spark_native_buckets_1uwewxkiauvk. Invalid method name: 'get_table_req';
spark               | 	at org.apache.spark.sql.hive.HiveExternalCatalog.withClient(HiveExternalCatalog.scala:109)
spark               | 	at org.apache.spark.sql.hive.HiveExternalCatalog.tableExists(HiveExternalCatalog.scala:851)
spark               | 	at org.apache.spark.sql.catalyst.catalog.ExternalCatalogWithListener.tableExists(ExternalCatalogWithListener.scala:146)
spark               | 	at org.apache.spark.sql.catalyst.catalog.SessionCatalog.tableExists(SessionCatalog.scala:432)
spark               | 	at org.apache.spark.sql.execution.command.CreateDataSourceTableCommand.run(createDataSourceTables.scala:52)
spark               | 	at org.apache.spark.sql.execution.command.ExecutedCommandExec.sideEffectResult$lzycompute(commands.scala:70)
spark               | 	at org.apache.spark.sql.execution.command.ExecutedCommandExec.sideEffectResult(commands.scala:68)
spark               | 	at org.apache.spark.sql.execution.command.ExecutedCommandExec.executeCollect(commands.scala:79)
spark               | 	at org.apache.spark.sql.Dataset.$anonfun$logicalPlan$1(Dataset.scala:229)
spark               | 	at org.apache.spark.sql.Dataset.$anonfun$withAction$1(Dataset.scala:3616)
spark               | 	at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$5(SQLExecution.scala:100)
spark               | 	at org.apache.spark.sql.execution.SQLExecution$.withSQLConfPropagated(SQLExecution.scala:160)
spark               | 	at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$1(SQLExecution.scala:87)
spark               | 	at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:763)
spark               | 	at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId(SQLExecution.scala:64)
spark               | 	at org.apache.spark.sql.Dataset.withAction(Dataset.scala:3614)
spark               | 	at org.apache.spark.sql.Dataset.<init>(Dataset.scala:229)
spark               | 	at org.apache.spark.sql.Dataset$.$anonfun$ofRows$2(Dataset.scala:100)
spark               | 	at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:763)
spark               | 	at org.apache.spark.sql.Dataset$.ofRows(Dataset.scala:97)
spark               | 	at org.apache.spark.sql.SparkSession.$anonfun$sql$1(SparkSession.scala:606)
spark               | 	at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:763)
spark               | 	at org.apache.spark.sql.SparkSession.sql(SparkSession.scala:601)
spark               | 	at org.apache.spark.sql.SQLContext.sql(SQLContext.scala:650)
spark               | 	at org.apache.spark.sql.hive.thriftserver.SparkExecuteStatementOperation.org$apache$spark$sql$hive$thriftserver$SparkExecuteStatementOperation$$execute(SparkExecuteStatementOperation.scala:280)
```

MiguelWeezardo commented:
@findepi Please merge if OK
